### PR TITLE
Fix #1130 - Can't run resstock in parallel

### LINKED ISSUE #1130: Can't run resstock in parallel
I know this is an odd workflow, but I'd like to call resstock in parallel, to run multiple models (baseline only) built each with a precomputed buildstock.csv

Issues:

* run_analysis.rb consistently wipes and recreates the resstock/lib folder
    * At initialization
        * cp `resstock/resources` to `resstock/lib/resources`
        * cp `resstock/project_<national|baseline>/housing_characteristics` to `resstock/lib/housing_characteristics`
    * Also wipes the lib folder at end unless `--debug` is passed
    * This is avoidable, especially considering there's a smart way to check if it's needed to recreate it, via checksuming the content of the source dir and target dir to ensure we have up to date copies in resstock/lib
    
* It copies (or create if not precomputed) the buildstock.csv to the `resstock/lib/housing_characteristics/buildstock.csv` and BuildExistingModel measure just plain assumes it there, with no option to change it
    * https://github.com/NREL/resstock/blob/1348def2ba52e446c35828e71f57cb146f2ac9be/measures/BuildExistingModel/measure.rb#L236

### PR DESCRIPTION
**Pull request by jmarrec**
## Pull Request Description

* Fix #1130 

I know the BuildExistingModel is in the OpenStudio-HPXML repo and changes should probably be made there, but I would like to make sure this has a chance of being included in ResStock before I make the changes over there as well.

Having the ability to run ResStock in parallel is something I actively need.

Running 125 "baseline" simulations on my 8 CPU / 16 cores machine:

* Serially: 40 minutes
* Parallel, with the changes here: 4min30 seconds.

I ran 3125 simulations, in paralle, in 2h30 minutes. And I plan to run a lot more.
This would have taken somewhere between 17 and 24 hours serially, rendering it completely unpractical

I'd very much like if this could be upstreamed in ResStock, as it may help other people and would reduce the burden of keeping my fork up to date.

## Checklist

Not all may apply:

- [ ] Tests (and test files) have been updated
- [ ] Documentation has been updated
  - [ ] If related to resstock-estimation, checklist includes [data dictionary](https://github.com/NREL/resstock/tree/develop/resources/data/dictionary), [source report](https://github.com/NREL/resstock/tree/develop/project_national/resources/source_report.csv), [options_lookup](https://github.com/NREL/resstock/blob/develop/resources/options_lookup.tsv).
  - [ ] If changes to project_testing tsvs, checklist includes [yml_precomputed](https://github.com/NREL/resstock/tree/develop/test/tests_yml_files/yml_precomputed), [yml_precomputed_outdated](https://github.com/NREL/resstock/tree/develop/test/tests_yml_files/yml_precomputed_outdated), [yml_precomputed_weight](https://github.com/NREL/resstock/tree/develop/test/tests_yml_files/yml_precomputed_weight)
- [ ] Changelog has been updated
- [ ] `openstudio tasks.rb update_measures` has been run
- [ ] No unexpected regression test changes on CI (checked comparison artifacts)

### CODE REVIEW

**Comment by joseph-robertson** [comment]
@jmarrec So the basic idea here is making the path of the buildstock.csv file variable (instead of fixed at lib/housing_characteristics) so that you can use `run_analysis.rb` in parallel?

Couple of questions.
(1) How can we get this branch running on CI?
(2) If we did, I'd assume we'd see problems with the integration-tests job that uses buildstockbatch (where you haven't yet updated its workflow generator for the BuildExistingModel update). Is there a way to make the new `buildstock_csv_path` argument optional (and default to the original lib/housing_characteristics) so we wouldn't _need_ to make any workflow generator changes?

**Comment by jmarrec** [comment]
> @jmarrec So the basic idea here is making the path of the buildstock.csv file variable (instead of fixed at lib/housing_characteristics) so that you can use `run_analysis.rb` in parallel?

Yes, and don't wipe the lib folder if it's already there and up to date.

> Couple of questions.

> (1) How can we get this branch running on CI? 

Your CI rules seem to protect against running on a branch from a fork... That's a bad setting IMHO. You should set it so workflows coming from a fork don't run automatically but required a maintainer to approve the run.

Anyways, I'm cloning this PR onto my forked repo so it runs there: https://github.com/jmarrec/resstock/pull/1


(2) If we did, I'd assume we'd see problems with the integration-tests job that uses buildstockbatch (where you haven't yet updated its workflow generator for the BuildExistingModel update). Is there a way to make the new `buildstock_csv_path` argument optional (and default to the original lib/housing_characteristics) so we wouldn't _need_ to make any workflow generator changes? Edit: looks like above you are already noting we could do this 👍 .

Yes we can. We'll see if it's necessary.

File: measures/BuildExistingModel/measure.rb [hunk 1]
```diff
@@ -30,6 +31,11 @@ def modeler_description
   def arguments(model) # rubocop:disable Lint/UnusedMethodArgument
     args = OpenStudio::Ruleset::OSArgumentVector.new
 
+    arg = OpenStudio::Ruleset::OSArgument.makeStringArgument('buildstock_csv_path', true)
```
Comment (jmarrec) [inline measures/BuildExistingModel/measure.rb:34]:
```suggestion
    arg = OpenStudio::Ruleset::OSArgument.makeStringArgument('buildstock_csv_path', false)
    arg.setDefaultValue("buildstock.csv")
```

**Comment by joseph-robertson** [comment]
@jmarrec Should we still consider getting this PR into a state to be merged?

**Comment by jmarrec** [comment]
@joseph-robertson I think the change makes sense, but as a maintainer you're always entitled to reject a pull request as something you don't want to maintain or don't see the point of. Up to you really. (I won't be mad if that's the case, just close it)

The conflicts are trivial to fix and I can resolve them if needed.

**Comment by joseph-robertson** [comment]
Superseded by https://github.com/NREL/resstock/pull/1230.

**Pull request closed without merging**